// ngx-schema-form/esm2015/lib/expression-compiler-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
export class ExpressionCompilerFactory {
}
if (false) {
    /**
     * @abstract
     * @return {?}
     */
    ExpressionCompilerFactory.prototype.createExpressionCompiler = function () { };
    /**
     * @abstract
     * @return {?}
     */
    ExpressionCompilerFactory.prototype.createExpressionCompilerVisibilityIf = function () { };
}
/**
 * @record
 */
export function ExpressionCompiler() { }
if (false) {
    /**
     * @param {?} expression
     * @param {?} context
     * @return {?}
     */
    ExpressionCompiler.prototype.evaluate = function (expression, context) { };
}
/**
 * @record
 */
export function ExpressionCompilerVisibilityIf() { }
if (false) {
    /**
     * @param {?} expression
     * @param {?} context
     * @return {?}
     */
    ExpressionCompilerVisibilityIf.prototype.evaluate = function (expression, context) { };
}
/**
 * UseCase:<br/>
 * When evaluating the expression of a <code>visibilityIf</code> condition
 * an instance of this definition will be passed as context.<br/>
 * This will give access to the source and target <code>FormProperty</code>.
 * @record
 */
export function ExpressionContextVisibilitIf() { }
if (false) {
    /**
     * The source property which has the <code>visibilityIf</code> defined
     * @type {?}
     */
    ExpressionContextVisibilitIf.prototype.source;
    /**
     * The target property given with the <code>visibilityIf</code>
     * <em>path</em> property
     * @type {?}
     */
    ExpressionContextVisibilitIf.prototype.target;
}
import * as JEXL from 'jexl';
export class JEXLExpressionCompilerFactory extends ExpressionCompilerFactory {
    /**
     * @return {?}
     */
    createExpressionCompiler() {
        return new JEXLExpressionCompiler();
    }
    /**
     * @return {?}
     */
    createExpressionCompilerVisibilityIf() {
        return new JEXLExpressionCompilerVisibiltyIf();
    }
}
export class JEXLExpressionCompiler {
    /**
     * @param {?} expression
     * @param {?=} context
     * @return {?}
     */
    evaluate(expression, context = {}) {
        return new JEXL.Jexl().evalSync(expression, context);
    }
}
export class JEXLExpressionCompilerVisibiltyIf {
    /**
     * @param {?} expression
     * @param {?=} context
     * @return {?}
     */
    evaluate(expression, context = { source: (/** @type {?} */ ({})), target: (/** @type {?} */ ({})) }) {
        return new JEXL.Jexl().evalSync(expression, context);
    }
}
