// ngx-schema-form/esm2015/lib/template-schema/field/item/item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input } from '@angular/core';
import { TemplateSchemaElement } from '../../template-schema-element';
export class ItemComponent extends TemplateSchemaElement {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        super();
        this.elementRef = elementRef;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.description = this.getTextContent(this.elementRef);
    }
}
ItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'sf-item',
                template: "<ng-content></ng-content>\n"
            }] }
];
/** @nocollapse */
ItemComponent.ctorParameters = () => [
    { type: ElementRef }
];
ItemComponent.propDecorators = {
    value: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    ItemComponent.prototype.value;
    /** @type {?} */
    ItemComponent.prototype.description;
    /**
     * @type {?}
     * @private
     */
    ItemComponent.prototype.elementRef;
}
