// @ngx-formly/primeng/esm2015/select/select.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: select.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { FormlyModule } from '@ngx-formly/core';
import { DropdownModule } from 'primeng/dropdown';
import { FormlySelectModule as FormlyCoreSelectModule } from '@ngx-formly/core/select';
import { FormlyFormFieldModule } from '@ngx-formly/primeng/form-field';
import { FormlyFieldSelect } from './select.type';
export class FormlySelectModule {
}
FormlySelectModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FormlyFieldSelect],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    DropdownModule,
                    FormlyFormFieldModule,
                    FormlyCoreSelectModule,
                    FormlyModule.forChild({
                        types: [
                            {
                                name: 'select',
                                component: FormlyFieldSelect,
                                wrappers: ['form-field'],
                            },
                        ],
                    }),
                ],
            },] }
];
