// ngx-schema-form/esm5/lib/model/formproperty.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BehaviorSubject, combineLatest } from 'rxjs';
import { distinctUntilChanged, map } from 'rxjs/operators';
/**
 * @abstract
 */
var /**
 * @abstract
 */
FormProperty = /** @class */ (function () {
    function FormProperty(schemaValidatorFactory, validatorRegistry, expressionCompilerFactory, schema, parent, path) {
        this.validatorRegistry = validatorRegistry;
        this.schema = schema;
        this._value = null;
        this._errors = null;
        this._valueChanges = new BehaviorSubject(null);
        this._errorsChanges = new BehaviorSubject(null);
        this._visible = true;
        this._visibilityChanges = new BehaviorSubject(true);
        this.schemaValidator = schemaValidatorFactory.createValidatorFn(this.schema);
        this.expressionCompilerVisibiltyIf = expressionCompilerFactory.createExpressionCompilerVisibilityIf();
        this._parent = parent;
        if (parent) {
            this._root = parent.root;
        }
        else if (this instanceof PropertyGroup) {
            this._root = (/** @type {?} */ ((/** @type {?} */ (this))));
            this._rootName = this.createRootName();
        }
        this._path = path;
    }
    Object.defineProperty(FormProperty.prototype, "_canonicalPath", {
        /**
         * Provides the unique path of this form element.<br/>
         * E.g.:
         * <code>/garage/cars</code>,<br/>
         * <code>/shop/book/0/page/1/</code>
         */
        get: /**
         * Provides the unique path of this form element.<br/>
         * E.g.:
         * <code>/garage/cars</code>,<br/>
         * <code>/shop/book/0/page/1/</code>
         * @return {?}
         */
        function () { return this.__canonicalPath; },
        set: /**
         * @param {?} canonicalPath
         * @return {?}
         */
        function (canonicalPath) {
            this.__canonicalPath = canonicalPath;
            this.__canonicalPathNotation = (this.__canonicalPath || '')
                .replace(new RegExp('^/', 'ig'), '')
                .replace(new RegExp('/$', 'ig'), '')
                .replace(new RegExp('/', 'ig'), '.');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "canonicalPathNotation", {
        /**
         * Uses the unique path provided by the property <code>_canonicalPath</code><br/>
         * but converts it to a HTML Element Attribute ID compliant format.<br/>
         * E.g.:
         * <code>garage.cars</code>,<br/>
         * <code>shop.book.0.page.1.</code>
         */
        get: /**
         * Uses the unique path provided by the property <code>_canonicalPath</code><br/>
         * but converts it to a HTML Element Attribute ID compliant format.<br/>
         * E.g.:
         * <code>garage.cars</code>,<br/>
         * <code>shop.book.0.page.1.</code>
         * @return {?}
         */
        function () { return this.__canonicalPathNotation; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "rootName", {
        /**
         * Provides the HTML Element Attribute ID/NAME compliant representation
         * of the root element.<br/>
         * Represents the HTML FORM NAME.<br/>
         * Only the root <code>FormProperty</code> will provide a value here.
         */
        get: /**
         * Provides the HTML Element Attribute ID/NAME compliant representation
         * of the root element.<br/>
         * Represents the HTML FORM NAME.<br/>
         * Only the root <code>FormProperty</code> will provide a value here.
         * @return {?}
         */
        function () { return this._rootName; },
        enumerable: true,
        configurable: true
    });
    /**
     * Creates the HTML ID and NAME attribute compliant string.
     */
    /**
     * Creates the HTML ID and NAME attribute compliant string.
     * @private
     * @return {?}
     */
    FormProperty.prototype.createRootName = /**
     * Creates the HTML ID and NAME attribute compliant string.
     * @private
     * @return {?}
     */
    function () {
        if (this.schema && this.schema['name']) {
            return this._rootName = this.schema['name'].replace(new RegExp('[\\s]+', 'ig'), '_');
        }
        return '';
    };
    Object.defineProperty(FormProperty.prototype, "valueChanges", {
        get: /**
         * @return {?}
         */
        function () {
            return this._valueChanges;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "errorsChanges", {
        get: /**
         * @return {?}
         */
        function () {
            return this._errorsChanges;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "type", {
        get: /**
         * @return {?}
         */
        function () {
            return this.schema.type;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "parent", {
        get: /**
         * @return {?}
         */
        function () {
            return this._parent;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "root", {
        get: /**
         * @return {?}
         */
        function () {
            return this._root || (/** @type {?} */ ((/** @type {?} */ (this))));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "path", {
        get: /**
         * @return {?}
         */
        function () {
            return this._path;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "visible", {
        get: /**
         * @return {?}
         */
        function () {
            return this._visible;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FormProperty.prototype, "valid", {
        get: /**
         * @return {?}
         */
        function () {
            return this._errors === null;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?=} onlySelf
     * @param {?=} emitEvent
     * @return {?}
     */
    FormProperty.prototype.updateValueAndValidity = /**
     * @param {?=} onlySelf
     * @param {?=} emitEvent
     * @return {?}
     */
    function (onlySelf, emitEvent) {
        if (onlySelf === void 0) { onlySelf = false; }
        if (emitEvent === void 0) { emitEvent = true; }
        this._updateValue();
        if (emitEvent) {
            this.valueChanges.next(this.value);
        }
        this._runValidation();
        if (this.parent && !onlySelf) {
            this.parent.updateValueAndValidity(onlySelf, emitEvent);
        }
    };
    /**
     * @internal
     */
    /**
     * \@internal
     * @return {?}
     */
    FormProperty.prototype._runValidation = /**
     * \@internal
     * @return {?}
     */
    function () {
        /** @type {?} */
        var errors = this.schemaValidator(this._value) || [];
        /** @type {?} */
        var customValidator = this.validatorRegistry.get(this.path);
        if (customValidator) {
            /** @type {?} */
            var customErrors = customValidator(this.value, this, this.findRoot());
            errors = this.mergeErrors(errors, customErrors);
        }
        if (errors.length === 0) {
            errors = null;
        }
        this._errors = errors;
        this.setErrors(this._errors);
    };
    /**
     * @private
     * @param {?} errors
     * @param {?} newErrors
     * @return {?}
     */
    FormProperty.prototype.mergeErrors = /**
     * @private
     * @param {?} errors
     * @param {?} newErrors
     * @return {?}
     */
    function (errors, newErrors) {
        if (newErrors) {
            if (Array.isArray(newErrors)) {
                errors = errors.concat.apply(errors, tslib_1.__spread(newErrors));
            }
            else {
                errors.push(newErrors);
            }
        }
        return errors;
    };
    /**
     * @private
     * @param {?} errors
     * @return {?}
     */
    FormProperty.prototype.setErrors = /**
     * @private
     * @param {?} errors
     * @return {?}
     */
    function (errors) {
        this._errors = errors;
        this._errorsChanges.next(errors);
    };
    /**
     * @param {?} errors
     * @return {?}
     */
    FormProperty.prototype.extendErrors = /**
     * @param {?} errors
     * @return {?}
     */
    function (errors) {
        errors = this.mergeErrors(this._errors || [], errors);
        this.setErrors(errors);
    };
    /**
     * @param {?} path
     * @return {?}
     */
    FormProperty.prototype.searchProperty = /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var prop = this;
        /** @type {?} */
        var base = null;
        /** @type {?} */
        var result = null;
        if (path[0] === '/') {
            base = this.findRoot();
            result = base.getProperty(path.substr(1));
        }
        else {
            while (result === null && prop.parent !== null) {
                prop = base = prop.parent;
                result = base.getProperty(path);
            }
        }
        return result;
    };
    /**
     * @return {?}
     */
    FormProperty.prototype.findRoot = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var property = this;
        while (property.parent !== null) {
            property = property.parent;
        }
        return (/** @type {?} */ (property));
    };
    /**
     * @private
     * @param {?} visible
     * @return {?}
     */
    FormProperty.prototype.setVisible = /**
     * @private
     * @param {?} visible
     * @return {?}
     */
    function (visible) {
        this._visible = visible;
        this._visibilityChanges.next(visible);
        this.updateValueAndValidity();
        if (this.parent) {
            this.parent.updateValueAndValidity(false, true);
        }
    };
    /**
     * Making use of the expression compiler for the <code>visibleIf</code> condition
     */
    /**
     * Making use of the expression compiler for the <code>visibleIf</code> condition
     * @private
     * @param {?} sourceProperty
     * @param {?} targetProperty
     * @param {?} dependencyPath
     * @param {?=} value
     * @param {?=} expression
     * @return {?}
     */
    FormProperty.prototype.__evaluateVisibilityIf = /**
     * Making use of the expression compiler for the <code>visibleIf</code> condition
     * @private
     * @param {?} sourceProperty
     * @param {?} targetProperty
     * @param {?} dependencyPath
     * @param {?=} value
     * @param {?=} expression
     * @return {?}
     */
    function (sourceProperty, targetProperty, dependencyPath, value, expression) {
        var e_1, _a;
        if (value === void 0) { value = ''; }
        if (expression === void 0) { expression = ''; }
        try {
            /** @type {?} */
            var valid = false;
            if (expression.indexOf('$ANY$') !== -1) {
                valid = value && value.length > 0;
            }
            else if ((expression || []).toString().indexOf('$EXP$') === 0) {
                // since visibleIf condition values are an array... we must do this
                /** @type {?} */
                var expArray = Array.isArray(expression) ? expression : (expression ? [expression] : []);
                try {
                    for (var expArray_1 = tslib_1.__values(expArray), expArray_1_1 = expArray_1.next(); !expArray_1_1.done; expArray_1_1 = expArray_1.next()) {
                        var expString = expArray_1_1.value;
                        /** @type {?} */
                        var _expresssion = expString.substring('$EXP$'.length);
                        valid = true === this.expressionCompilerVisibiltyIf.evaluate(_expresssion, {
                            source: sourceProperty,
                            target: targetProperty
                        });
                        if (valid) {
                            break;
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (expArray_1_1 && !expArray_1_1.done && (_a = expArray_1.return)) _a.call(expArray_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            else {
                valid = expression.indexOf(value) !== -1;
            }
            return valid;
        }
        catch (error) {
            console.error('Error processing "VisibileIf" expression for path: ', dependencyPath, "source - " + sourceProperty._canonicalPath + ": ", sourceProperty, "target - " + targetProperty._canonicalPath + ": ", targetProperty, 'value:', value, 'expression: ', expression, 'error: ', error);
        }
    };
    /**
     * @private
     * @return {?}
     */
    FormProperty.prototype.__bindVisibility = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        var e_2, _a;
        /**
         * <pre>
         *     "oneOf":[{
         *         "path":["value","value"]
         *     },{
         *         "path":["value","value"]
         *     }]
         *     </pre>
         * <pre>
         *     "allOf":[{
         *         "path":["value","value"]
         *     },{
         *         "path":["value","value"]
         *     }]
         *     </pre>
         * @type {?}
         */
        var visibleIfProperty = this.schema.visibleIf;
        /** @type {?} */
        var visibleIfOf = (visibleIfProperty || {}).oneOf || (visibleIfProperty || {}).allOf;
        if (visibleIfOf) {
            var _loop_1 = function (visibleIf) {
                if (typeof visibleIf === 'object' && Object.keys(visibleIf).length === 0) {
                    this_1.setVisible(false);
                }
                else if (visibleIf !== undefined) {
                    /** @type {?} */
                    var propertiesBinding = [];
                    var _loop_2 = function (dependencyPath) {
                        var e_3, _a;
                        if (visibleIf.hasOwnProperty(dependencyPath)) {
                            /** @type {?} */
                            var properties = this_1.findProperties(this_1, dependencyPath);
                            if ((properties || []).length) {
                                var _loop_3 = function (property) {
                                    if (property) {
                                        /** @type {?} */
                                        var valueCheck = void 0;
                                        if (this_1.schema.visibleIf.oneOf) {
                                            valueCheck = property.valueChanges.pipe(map((/**
                                             * @param {?} value
                                             * @return {?}
                                             */
                                            function (value) { return _this.__evaluateVisibilityIf(_this, property, dependencyPath, value, visibleIf[dependencyPath]); })));
                                        }
                                        else if (this_1.schema.visibleIf.allOf) {
                                            /** @type {?} */
                                            var _chk = (/**
                                             * @param {?} value
                                             * @return {?}
                                             */
                                            function (value) {
                                                var e_4, _a, e_5, _b;
                                                try {
                                                    for (var _c = tslib_1.__values(_this.schema.visibleIf.allOf), _d = _c.next(); !_d.done; _d = _c.next()) {
                                                        var item = _d.value;
                                                        try {
                                                            for (var _e = tslib_1.__values(Object.keys(item)), _f = _e.next(); !_f.done; _f = _e.next()) {
                                                                var depPath = _f.value;
                                                                /** @type {?} */
                                                                var prop = _this.searchProperty(depPath);
                                                                /** @type {?} */
                                                                var propVal = prop.value;
                                                                if (!_this.__evaluateVisibilityIf(_this, prop, dependencyPath, propVal, item[depPath])) {
                                                                    return false;
                                                                }
                                                            }
                                                        }
                                                        catch (e_5_1) { e_5 = { error: e_5_1 }; }
                                                        finally {
                                                            try {
                                                                if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                                                            }
                                                            finally { if (e_5) throw e_5.error; }
                                                        }
                                                    }
                                                }
                                                catch (e_4_1) { e_4 = { error: e_4_1 }; }
                                                finally {
                                                    try {
                                                        if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                                                    }
                                                    finally { if (e_4) throw e_4.error; }
                                                }
                                                return true;
                                            });
                                            valueCheck = property.valueChanges.pipe(map(_chk));
                                        }
                                        /** @type {?} */
                                        var visibilityCheck = property._visibilityChanges;
                                        /** @type {?} */
                                        var and = combineLatest([valueCheck, visibilityCheck], (/**
                                         * @param {?} v1
                                         * @param {?} v2
                                         * @return {?}
                                         */
                                        function (v1, v2) { return v1 && v2; }));
                                        propertiesBinding.push(and);
                                    }
                                };
                                try {
                                    for (var properties_1 = tslib_1.__values(properties), properties_1_1 = properties_1.next(); !properties_1_1.done; properties_1_1 = properties_1.next()) {
                                        var property = properties_1_1.value;
                                        _loop_3(property);
                                    }
                                }
                                catch (e_3_1) { e_3 = { error: e_3_1 }; }
                                finally {
                                    try {
                                        if (properties_1_1 && !properties_1_1.done && (_a = properties_1.return)) _a.call(properties_1);
                                    }
                                    finally { if (e_3) throw e_3.error; }
                                }
                            }
                            else {
                                console.warn('Can\'t find property ' + dependencyPath + ' for visibility check of ' + this_1.path);
                                this_1.registerMissingVisibilityBinding(dependencyPath, this_1);
                                // not visible if not existent
                                this_1.setVisible(false);
                            }
                        }
                    };
                    for (var dependencyPath in visibleIf) {
                        _loop_2(dependencyPath);
                    }
                    combineLatest(propertiesBinding, (/**
                     * @param {...?} values
                     * @return {?}
                     */
                    function () {
                        var values = [];
                        for (var _i = 0; _i < arguments.length; _i++) {
                            values[_i] = arguments[_i];
                        }
                        return values.indexOf(true) !== -1;
                    })).pipe(distinctUntilChanged()).subscribe((/**
                     * @param {?} visible
                     * @return {?}
                     */
                    function (visible) {
                        _this.setVisible(visible);
                    }));
                }
            };
            var this_1 = this;
            try {
                for (var visibleIfOf_1 = tslib_1.__values(visibleIfOf), visibleIfOf_1_1 = visibleIfOf_1.next(); !visibleIfOf_1_1.done; visibleIfOf_1_1 = visibleIfOf_1.next()) {
                    var visibleIf = visibleIfOf_1_1.value;
                    _loop_1(visibleIf);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (visibleIfOf_1_1 && !visibleIfOf_1_1.done && (_a = visibleIfOf_1.return)) _a.call(visibleIfOf_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return true;
        }
    };
    // A field is visible if AT LEAST ONE of the properties it depends on is visible AND has a value in the list
    // A field is visible if AT LEAST ONE of the properties it depends on is visible AND has a value in the list
    /**
     * @return {?}
     */
    FormProperty.prototype._bindVisibility = 
    // A field is visible if AT LEAST ONE of the properties it depends on is visible AND has a value in the list
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.__bindVisibility())
            return;
        /** @type {?} */
        var visibleIf = this.schema.visibleIf;
        if (typeof visibleIf === 'object' && Object.keys(visibleIf).length === 0) {
            this.setVisible(false);
        }
        else if (visibleIf !== undefined) {
            /** @type {?} */
            var propertiesBinding = [];
            var _loop_4 = function (dependencyPath) {
                var e_6, _a;
                if (visibleIf.hasOwnProperty(dependencyPath)) {
                    /** @type {?} */
                    var properties = this_2.findProperties(this_2, dependencyPath);
                    if ((properties || []).length) {
                        var _loop_5 = function (property) {
                            if (property) {
                                /** @type {?} */
                                var valueCheck = property.valueChanges.pipe(map((/**
                                 * @param {?} value
                                 * @return {?}
                                 */
                                function (value) { return _this.__evaluateVisibilityIf(_this, property, dependencyPath, value, visibleIf[dependencyPath]); })));
                                /** @type {?} */
                                var visibilityCheck = property._visibilityChanges;
                                /** @type {?} */
                                var and = combineLatest([valueCheck, visibilityCheck], (/**
                                 * @param {?} v1
                                 * @param {?} v2
                                 * @return {?}
                                 */
                                function (v1, v2) { return v1 && v2; }));
                                propertiesBinding.push(and);
                            }
                        };
                        try {
                            for (var properties_2 = tslib_1.__values(properties), properties_2_1 = properties_2.next(); !properties_2_1.done; properties_2_1 = properties_2.next()) {
                                var property = properties_2_1.value;
                                _loop_5(property);
                            }
                        }
                        catch (e_6_1) { e_6 = { error: e_6_1 }; }
                        finally {
                            try {
                                if (properties_2_1 && !properties_2_1.done && (_a = properties_2.return)) _a.call(properties_2);
                            }
                            finally { if (e_6) throw e_6.error; }
                        }
                    }
                    else {
                        console.warn('Can\'t find property ' + dependencyPath + ' for visibility check of ' + this_2.path);
                        this_2.registerMissingVisibilityBinding(dependencyPath, this_2);
                        // not visible if not existent
                        this_2.setVisible(false);
                    }
                }
            };
            var this_2 = this;
            for (var dependencyPath in visibleIf) {
                _loop_4(dependencyPath);
            }
            combineLatest(propertiesBinding, (/**
             * @param {...?} values
             * @return {?}
             */
            function () {
                var values = [];
                for (var _i = 0; _i < arguments.length; _i++) {
                    values[_i] = arguments[_i];
                }
                return values.indexOf(true) !== -1;
            })).pipe(distinctUntilChanged()).subscribe((/**
             * @param {?} visible
             * @return {?}
             */
            function (visible) {
                _this.setVisible(visible);
            }));
        }
    };
    /**
     * @private
     * @param {?} dependencyPath
     * @param {?} formProperty
     * @return {?}
     */
    FormProperty.prototype.registerMissingVisibilityBinding = /**
     * @private
     * @param {?} dependencyPath
     * @param {?} formProperty
     * @return {?}
     */
    function (dependencyPath, formProperty) {
        formProperty._propertyBindingRegistry.getPropertyBindingsVisibility().add(dependencyPath, formProperty.path);
    };
    /**
     * Finds all <code>formProperties</code> from a path with wildcards.<br/>
     * e.g: <code>/garage/cars/&#42;/tires/&#42;/name</code><br/>
     * @param target
     * @param propertyPath
     */
    /**
     * Finds all <code>formProperties</code> from a path with wildcards.<br/>
     * e.g: <code>/garage/cars/&#42;/tires/&#42;/name</code><br/>
     * @param {?} target
     * @param {?} propertyPath
     * @return {?}
     */
    FormProperty.prototype.findProperties = /**
     * Finds all <code>formProperties</code> from a path with wildcards.<br/>
     * e.g: <code>/garage/cars/&#42;/tires/&#42;/name</code><br/>
     * @param {?} target
     * @param {?} propertyPath
     * @return {?}
     */
    function (target, propertyPath) {
        var e_7, _a;
        /** @type {?} */
        var props = [];
        /** @type {?} */
        var paths = this.findPropertyPaths(target, propertyPath);
        try {
            for (var paths_1 = tslib_1.__values(paths), paths_1_1 = paths_1.next(); !paths_1_1.done; paths_1_1 = paths_1.next()) {
                var path = paths_1_1.value;
                /** @type {?} */
                var p = target.searchProperty(path);
                if (p) {
                    props.push(p);
                }
            }
        }
        catch (e_7_1) { e_7 = { error: e_7_1 }; }
        finally {
            try {
                if (paths_1_1 && !paths_1_1.done && (_a = paths_1.return)) _a.call(paths_1);
            }
            finally { if (e_7) throw e_7.error; }
        }
        return props;
    };
    /**
     * Creates canonical paths from a path with wildcards.
     * e.g:<br/>
     * From:<br/>
     * <code>/garage/cars/&#42;/tires/&#42;/name</code><br/>
     * it creates:<br/>
     * <code>/garage/cars/0/tires/0/name</code><br/>
     * <code>/garage/cars/0/tires/1/name</code><br/>
     * <code>/garage/cars/0/tires/2/name</code><br/>
     * <code>/garage/cars/0/tires/3/name</code><br/>
     * <code>/garage/cars/1/tires/0/name</code><br/>
     * <code>/garage/cars/2/tires/1/name</code><br/>
     * <code>/garage/cars/3/tires/2/name</code><br/>
     * <code>/garage/cars/3/tires/3/name</code><br/>
     * <code>/garage/cars/&#42;/tires/&#42;/name</code><br/>
     * <code>/garage/cars/&#42;/tires/2/name</code><br/>
     * <code>/garage/cars/&#42;/tires/3/name</code><br/>
     * <br/>etc...
     * @param target
     * @param path
     * @param parentPath
     */
    /**
     * Creates canonical paths from a path with wildcards.
     * e.g:<br/>
     * From:<br/>
     * <code>/garage/cars/&#42;/tires/&#42;/name</code><br/>
     * it creates:<br/>
     * <code>/garage/cars/0/tires/0/name</code><br/>
     * <code>/garage/cars/0/tires/1/name</code><br/>
     * <code>/garage/cars/0/tires/2/name</code><br/>
     * <code>/garage/cars/0/tires/3/name</code><br/>
     * <code>/garage/cars/1/tires/0/name</code><br/>
     * <code>/garage/cars/2/tires/1/name</code><br/>
     * <code>/garage/cars/3/tires/2/name</code><br/>
     * <code>/garage/cars/3/tires/3/name</code><br/>
     * <code>/garage/cars/&#42;/tires/&#42;/name</code><br/>
     * <code>/garage/cars/&#42;/tires/2/name</code><br/>
     * <code>/garage/cars/&#42;/tires/3/name</code><br/>
     * <br/>etc...
     * @param {?} target
     * @param {?} path
     * @param {?=} parentPath
     * @return {?}
     */
    FormProperty.prototype.findPropertyPaths = /**
     * Creates canonical paths from a path with wildcards.
     * e.g:<br/>
     * From:<br/>
     * <code>/garage/cars/&#42;/tires/&#42;/name</code><br/>
     * it creates:<br/>
     * <code>/garage/cars/0/tires/0/name</code><br/>
     * <code>/garage/cars/0/tires/1/name</code><br/>
     * <code>/garage/cars/0/tires/2/name</code><br/>
     * <code>/garage/cars/0/tires/3/name</code><br/>
     * <code>/garage/cars/1/tires/0/name</code><br/>
     * <code>/garage/cars/2/tires/1/name</code><br/>
     * <code>/garage/cars/3/tires/2/name</code><br/>
     * <code>/garage/cars/3/tires/3/name</code><br/>
     * <code>/garage/cars/&#42;/tires/&#42;/name</code><br/>
     * <code>/garage/cars/&#42;/tires/2/name</code><br/>
     * <code>/garage/cars/&#42;/tires/3/name</code><br/>
     * <br/>etc...
     * @param {?} target
     * @param {?} path
     * @param {?=} parentPath
     * @return {?}
     */
    function (target, path, parentPath) {
        /** @type {?} */
        var ix = path.indexOf('*');
        if (-1 !== ix) {
            /** @type {?} */
            var prePath = ix > -1 ? path.substring(0, ix - 1) : path;
            /** @type {?} */
            var subPath = ix > -1 ? path.substring(ix + 1) : path;
            /** @type {?} */
            var prop = target.searchProperty(prePath);
            /** @type {?} */
            var pathFound = [];
            if (prop instanceof PropertyGroup) {
                /** @type {?} */
                var arrProp = (/** @type {?} */ (prop.properties));
                for (var i = 0; i < arrProp.length; i++) {
                    /** @type {?} */
                    var curreItemPath = (parentPath || '') + prePath + (prePath.endsWith('/') ? '' : '/') + i + subPath;
                    /** @type {?} */
                    var curreItemPrePath = (parentPath || '') + prePath + i;
                    if (-1 === curreItemPath.indexOf('*')) {
                        pathFound.push(curreItemPath);
                    }
                    /** @type {?} */
                    var childrenPathFound = this.findPropertyPaths(arrProp[i], subPath, curreItemPrePath);
                    pathFound = pathFound.concat(childrenPathFound);
                }
            }
            return pathFound;
        }
        return [path];
    };
    return FormProperty;
}());
/**
 * @abstract
 */
export { FormProperty };
if (false) {
    /** @type {?} */
    FormProperty.prototype.schemaValidator;
    /** @type {?} */
    FormProperty.prototype.expressionCompilerVisibiltyIf;
    /** @type {?} */
    FormProperty.prototype._value;
    /** @type {?} */
    FormProperty.prototype._errors;
    /**
     * @type {?}
     * @private
     */
    FormProperty.prototype._valueChanges;
    /**
     * @type {?}
     * @private
     */
    FormProperty.prototype._errorsChanges;
    /**
     * @type {?}
     * @private
     */
    FormProperty.prototype._visible;
    /**
     * @type {?}
     * @private
     */
    FormProperty.prototype._visibilityChanges;
    /**
     * @type {?}
     * @private
     */
    FormProperty.prototype._root;
    /**
     * @type {?}
     * @private
     */
    FormProperty.prototype._parent;
    /**
     * @type {?}
     * @private
     */
    FormProperty.prototype._path;
    /** @type {?} */
    FormProperty.prototype._propertyBindingRegistry;
    /** @type {?} */
    FormProperty.prototype.__canonicalPath;
    /** @type {?} */
    FormProperty.prototype.__canonicalPathNotation;
    /**
     * @type {?}
     * @private
     */
    FormProperty.prototype._rootName;
    /**
     * @type {?}
     * @private
     */
    FormProperty.prototype.validatorRegistry;
    /** @type {?} */
    FormProperty.prototype.schema;
    /**
     * @abstract
     * @param {?} value
     * @param {?} onlySelf
     * @return {?}
     */
    FormProperty.prototype.setValue = function (value, onlySelf) { };
    /**
     * @abstract
     * @param {?} value
     * @param {?} onlySelf
     * @return {?}
     */
    FormProperty.prototype.reset = function (value, onlySelf) { };
    /**
     * \@internal
     * @abstract
     * @return {?}
     */
    FormProperty.prototype._hasValue = function () { };
    /**
     * \@internal
     * @abstract
     * @return {?}
     */
    FormProperty.prototype._updateValue = function () { };
}
/**
 * @abstract
 */
var /**
 * @abstract
 */
PropertyGroup = /** @class */ (function (_super) {
    tslib_1.__extends(PropertyGroup, _super);
    function PropertyGroup() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._properties = null;
        _this._propertyProxyHandler = {
            /**
             * When a new item is added it will be checked for visibility updates to proceed <br/>
             * if any other field has a binding reference to it.<br/>
             */
            set: /**
             * When a new item is added it will be checked for visibility updates to proceed <br/>
             * if any other field has a binding reference to it.<br/>
             * @param {?} target
             * @param {?} p
             * @param {?} value
             * @param {?} receiver
             * @return {?}
             */
            function (target, p, value, receiver) {
                /**
                 * 1) Make sure a canonical path is set
                 * @type {?}
                 */
                var assertCanonicalPath = (/**
                 * @param {?} propertyValue
                 * @return {?}
                 */
                function (propertyValue) {
                    var e_8, _a;
                    /** @type {?} */
                    var formProperty = (/** @type {?} */ (propertyValue));
                    if (Array.isArray(target) && propertyValue instanceof FormProperty) {
                        /**
                         * Create a canonical path replacing the last '*' with the elements position in array
                         * \@param propertyPath
                         * \@param indexOfChild
                         * @type {?}
                         */
                        var getCanonicalPath = (/**
                         * @param {?} propertyPath
                         * @param {?} indexOfChild
                         * @return {?}
                         */
                        function (propertyPath, indexOfChild) {
                            /** @type {?} */
                            var pos;
                            if (propertyPath && -1 !== (pos = propertyPath.lastIndexOf('*'))) {
                                return propertyPath.substring(0, pos) + indexOfChild.toString() + propertyPath.substring(pos + 1);
                            }
                        });
                        if (formProperty) {
                            formProperty._canonicalPath = getCanonicalPath(formProperty._canonicalPath, (/** @type {?} */ (p)));
                        }
                    }
                    /** @type {?} */
                    var propertyGroup = (/** @type {?} */ (formProperty));
                    /** @type {?} */
                    var propertyGroupChildren = (/** @type {?} */ ((Array.isArray(propertyGroup.properties) ?
                        propertyGroup.properties :
                        Object.values(propertyGroup.properties || {}))));
                    if ((formProperty.path || '').endsWith('/*')) {
                        try {
                            /**
                             * If it is an array, then all children canonical paths must be computed now.
                             * The children don't have the parent's path segment set yet,
                             * because they are created before the parent gets attached to its parent.
                             */
                            for (var propertyGroupChildren_1 = tslib_1.__values(propertyGroupChildren), propertyGroupChildren_1_1 = propertyGroupChildren_1.next(); !propertyGroupChildren_1_1.done; propertyGroupChildren_1_1 = propertyGroupChildren_1.next()) {
                                var child = propertyGroupChildren_1_1.value;
                                child._canonicalPath = formProperty._canonicalPath + child._canonicalPath.substring(formProperty.path.length);
                            }
                        }
                        catch (e_8_1) { e_8 = { error: e_8_1 }; }
                        finally {
                            try {
                                if (propertyGroupChildren_1_1 && !propertyGroupChildren_1_1.done && (_a = propertyGroupChildren_1.return)) _a.call(propertyGroupChildren_1);
                            }
                            finally { if (e_8) throw e_8.error; }
                        }
                    }
                    return { property: formProperty, children: propertyGroupChildren };
                });
                var _a = assertCanonicalPath(value), property = _a.property, children = _a.children;
                /**
                 * 2) Add the new property before rebinding, so it can be found by <code>_bindVisibility</code>
                 * @type {?}
                 */
                var result = target[(/** @type {?} */ (p))] = value;
                /**
                 * 3) Re-bind the visibility bindings referencing to this canonical paths
                 * @type {?}
                 */
                var rebindVisibility = (/**
                 * @return {?}
                 */
                function () {
                    var e_9, _a, e_10, _b;
                    /** @type {?} */
                    var rebindAll = [property].concat(children);
                    /** @type {?} */
                    var findPropertiesToRebind = (/**
                     * @param {?} formProperty
                     * @return {?}
                     */
                    function (formProperty) {
                        var e_11, _a;
                        /** @type {?} */
                        var propertyBindings = formProperty._propertyBindingRegistry.getPropertyBindingsVisibility();
                        /** @type {?} */
                        var rebind = [];
                        if (formProperty._canonicalPath) {
                            rebind = rebind.concat(rebind.concat(propertyBindings.findByDependencyPath(formProperty._canonicalPath) || []));
                            if (formProperty._canonicalPath.startsWith('/')) {
                                rebind = rebind.concat(rebind.concat(propertyBindings.findByDependencyPath(formProperty._canonicalPath.substring(1)) || []));
                            }
                        }
                        rebind = rebind.concat(propertyBindings.findByDependencyPath(formProperty.path) || []);
                        if (formProperty.path.startsWith('/')) {
                            rebind = rebind.concat(rebind.concat(propertyBindings.findByDependencyPath(formProperty.path.substring(1)) || []));
                        }
                        /** @type {?} */
                        var uniqueValues = {};
                        try {
                            for (var rebind_1 = tslib_1.__values(rebind), rebind_1_1 = rebind_1.next(); !rebind_1_1.done; rebind_1_1 = rebind_1.next()) {
                                var item = rebind_1_1.value;
                                uniqueValues[item] = item;
                            }
                        }
                        catch (e_11_1) { e_11 = { error: e_11_1 }; }
                        finally {
                            try {
                                if (rebind_1_1 && !rebind_1_1.done && (_a = rebind_1.return)) _a.call(rebind_1);
                            }
                            finally { if (e_11) throw e_11.error; }
                        }
                        return Object.keys(uniqueValues);
                    });
                    try {
                        for (var rebindAll_1 = tslib_1.__values(rebindAll), rebindAll_1_1 = rebindAll_1.next(); !rebindAll_1_1.done; rebindAll_1_1 = rebindAll_1.next()) {
                            var _property = rebindAll_1_1.value;
                            if (_property instanceof FormProperty) {
                                try {
                                    /** @type {?} */
                                    var rebindPaths = findPropertiesToRebind(_property);
                                    try {
                                        for (var rebindPaths_1 = tslib_1.__values(rebindPaths), rebindPaths_1_1 = rebindPaths_1.next(); !rebindPaths_1_1.done; rebindPaths_1_1 = rebindPaths_1.next()) {
                                            var rebindPropPath = rebindPaths_1_1.value;
                                            /** @type {?} */
                                            var rebindProp = _property.searchProperty(rebindPropPath);
                                            rebindProp._bindVisibility();
                                        }
                                    }
                                    catch (e_10_1) { e_10 = { error: e_10_1 }; }
                                    finally {
                                        try {
                                            if (rebindPaths_1_1 && !rebindPaths_1_1.done && (_b = rebindPaths_1.return)) _b.call(rebindPaths_1);
                                        }
                                        finally { if (e_10) throw e_10.error; }
                                    }
                                }
                                catch (e) {
                                    console.error('Rebinding visibility error at path:', _property.path, 'property:', _property, e);
                                }
                            }
                        }
                    }
                    catch (e_9_1) { e_9 = { error: e_9_1 }; }
                    finally {
                        try {
                            if (rebindAll_1_1 && !rebindAll_1_1.done && (_a = rebindAll_1.return)) _a.call(rebindAll_1);
                        }
                        finally { if (e_9) throw e_9.error; }
                    }
                });
                rebindVisibility();
                return result;
            },
            get: /**
             * @param {?} target
             * @param {?} p
             * @param {?} receiver
             * @return {?}
             */
            function (target, p, receiver) {
                return target[(/** @type {?} */ (p))];
            },
            deleteProperty: /**
             * @param {?} target
             * @param {?} p
             * @return {?}
             */
            function (target, p) {
                return delete target[(/** @type {?} */ (p))];
            }
        };
        return _this;
    }
    Object.defineProperty(PropertyGroup.prototype, "properties", {
        get: /**
         * @return {?}
         */
        function () {
            return this._properties;
        },
        set: /**
         * @param {?} properties
         * @return {?}
         */
        function (properties) {
            /**
             * Override the setter to add an observer that notices when an item is added or removed.<br/>
             */
            this._properties = new Proxy(properties, this._propertyProxyHandler);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} path
     * @return {?}
     */
    PropertyGroup.prototype.getProperty = /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var subPathIdx = path.indexOf('/');
        /** @type {?} */
        var propertyId = subPathIdx !== -1 ? path.substr(0, subPathIdx) : path;
        /** @type {?} */
        var property = this.properties[propertyId];
        if (property !== null && subPathIdx !== -1 && property instanceof PropertyGroup) {
            /** @type {?} */
            var subPath = path.substr(subPathIdx + 1);
            property = ((/** @type {?} */ (property))).getProperty(subPath);
        }
        return property;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    PropertyGroup.prototype.forEachChild = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        for (var propertyId in this.properties) {
            if (this.properties.hasOwnProperty(propertyId)) {
                /** @type {?} */
                var property = this.properties[propertyId];
                fn(property, propertyId);
            }
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    PropertyGroup.prototype.forEachChildRecursive = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.forEachChild((/**
         * @param {?} child
         * @return {?}
         */
        function (child) {
            fn(child);
            if (child instanceof PropertyGroup) {
                ((/** @type {?} */ (child))).forEachChildRecursive(fn);
            }
        }));
    };
    /**
     * @return {?}
     */
    PropertyGroup.prototype._bindVisibility = /**
     * @return {?}
     */
    function () {
        _super.prototype._bindVisibility.call(this);
        this._bindVisibilityRecursive();
    };
    /**
     * @private
     * @return {?}
     */
    PropertyGroup.prototype._bindVisibilityRecursive = /**
     * @private
     * @return {?}
     */
    function () {
        this.forEachChildRecursive((/**
         * @param {?} property
         * @return {?}
         */
        function (property) {
            property._bindVisibility();
        }));
    };
    /**
     * @return {?}
     */
    PropertyGroup.prototype.isRoot = /**
     * @return {?}
     */
    function () {
        return this === this.root;
    };
    return PropertyGroup;
}(FormProperty));
/**
 * @abstract
 */
export { PropertyGroup };
if (false) {
    /** @type {?} */
    PropertyGroup.prototype._properties;
    /**
     * @type {?}
     * @private
     */
    PropertyGroup.prototype._propertyProxyHandler;
}
