// ngx-schema-form/esm5/lib/expression-compiler-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/**
 * @abstract
 */
var /**
 * @abstract
 */
ExpressionCompilerFactory = /** @class */ (function () {
    function ExpressionCompilerFactory() {
    }
    return ExpressionCompilerFactory;
}());
/**
 * @abstract
 */
export { ExpressionCompilerFactory };
if (false) {
    /**
     * @abstract
     * @return {?}
     */
    ExpressionCompilerFactory.prototype.createExpressionCompiler = function () { };
    /**
     * @abstract
     * @return {?}
     */
    ExpressionCompilerFactory.prototype.createExpressionCompilerVisibilityIf = function () { };
}
/**
 * @record
 */
export function ExpressionCompiler() { }
if (false) {
    /**
     * @param {?} expression
     * @param {?} context
     * @return {?}
     */
    ExpressionCompiler.prototype.evaluate = function (expression, context) { };
}
/**
 * @record
 */
export function ExpressionCompilerVisibilityIf() { }
if (false) {
    /**
     * @param {?} expression
     * @param {?} context
     * @return {?}
     */
    ExpressionCompilerVisibilityIf.prototype.evaluate = function (expression, context) { };
}
/**
 * UseCase:<br/>
 * When evaluating the expression of a <code>visibilityIf</code> condition
 * an instance of this definition will be passed as context.<br/>
 * This will give access to the source and target <code>FormProperty</code>.
 * @record
 */
export function ExpressionContextVisibilitIf() { }
if (false) {
    /**
     * The source property which has the <code>visibilityIf</code> defined
     * @type {?}
     */
    ExpressionContextVisibilitIf.prototype.source;
    /**
     * The target property given with the <code>visibilityIf</code>
     * <em>path</em> property
     * @type {?}
     */
    ExpressionContextVisibilitIf.prototype.target;
}
import * as JEXL from 'jexl';
var JEXLExpressionCompilerFactory = /** @class */ (function (_super) {
    tslib_1.__extends(JEXLExpressionCompilerFactory, _super);
    function JEXLExpressionCompilerFactory() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    JEXLExpressionCompilerFactory.prototype.createExpressionCompiler = /**
     * @return {?}
     */
    function () {
        return new JEXLExpressionCompiler();
    };
    /**
     * @return {?}
     */
    JEXLExpressionCompilerFactory.prototype.createExpressionCompilerVisibilityIf = /**
     * @return {?}
     */
    function () {
        return new JEXLExpressionCompilerVisibiltyIf();
    };
    return JEXLExpressionCompilerFactory;
}(ExpressionCompilerFactory));
export { JEXLExpressionCompilerFactory };
var JEXLExpressionCompiler = /** @class */ (function () {
    function JEXLExpressionCompiler() {
    }
    /**
     * @param {?} expression
     * @param {?=} context
     * @return {?}
     */
    JEXLExpressionCompiler.prototype.evaluate = /**
     * @param {?} expression
     * @param {?=} context
     * @return {?}
     */
    function (expression, context) {
        if (context === void 0) { context = {}; }
        return new JEXL.Jexl().evalSync(expression, context);
    };
    return JEXLExpressionCompiler;
}());
export { JEXLExpressionCompiler };
var JEXLExpressionCompilerVisibiltyIf = /** @class */ (function () {
    function JEXLExpressionCompilerVisibiltyIf() {
    }
    /**
     * @param {?} expression
     * @param {?=} context
     * @return {?}
     */
    JEXLExpressionCompilerVisibiltyIf.prototype.evaluate = /**
     * @param {?} expression
     * @param {?=} context
     * @return {?}
     */
    function (expression, context) {
        if (context === void 0) { context = { source: (/** @type {?} */ ({})), target: (/** @type {?} */ ({})) }; }
        return new JEXL.Jexl().evalSync(expression, context);
    };
    return JEXLExpressionCompilerVisibiltyIf;
}());
export { JEXLExpressionCompilerVisibiltyIf };
