// ngx-schema-form/esm5/lib/formelement.action.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, ViewContainerRef } from "@angular/core";
import { WidgetFactory } from "./widgetfactory";
import { TerminatorService } from "./terminator.service";
var FormElementComponentAction = /** @class */ (function () {
    function FormElementComponentAction(widgetFactory, terminator) {
        if (widgetFactory === void 0) { widgetFactory = null; }
        this.widgetFactory = widgetFactory;
        this.terminator = terminator;
    }
    /**
     * @return {?}
     */
    FormElementComponentAction.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.subs = this.terminator.onDestroy.subscribe((/**
         * @param {?} destroy
         * @return {?}
         */
        function (destroy) {
            if (destroy) {
                _this.ref.destroy();
            }
        }));
    };
    /**
     * @return {?}
     */
    FormElementComponentAction.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.ref = this.widgetFactory.createWidget(this.container, this.button.widget || 'button');
        this.ref.instance.button = this.button;
        this.ref.instance.formProperty = this.formProperty;
    };
    /**
     * @return {?}
     */
    FormElementComponentAction.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subs.unsubscribe();
    };
    FormElementComponentAction.decorators = [
        { type: Component, args: [{
                    selector: 'sf-form-element-action',
                    template: '<ng-template #target></ng-template>'
                }] }
    ];
    /** @nocollapse */
    FormElementComponentAction.ctorParameters = function () { return [
        { type: WidgetFactory },
        { type: TerminatorService }
    ]; };
    FormElementComponentAction.propDecorators = {
        button: [{ type: Input }],
        formProperty: [{ type: Input }],
        container: [{ type: ViewChild, args: ['target', { read: ViewContainerRef, static: true },] }]
    };
    return FormElementComponentAction;
}());
export { FormElementComponentAction };
if (false) {
    /** @type {?} */
    FormElementComponentAction.prototype.button;
    /** @type {?} */
    FormElementComponentAction.prototype.formProperty;
    /** @type {?} */
    FormElementComponentAction.prototype.container;
    /**
     * @type {?}
     * @private
     */
    FormElementComponentAction.prototype.ref;
    /**
     * @type {?}
     * @private
     */
    FormElementComponentAction.prototype.subs;
    /**
     * @type {?}
     * @private
     */
    FormElementComponentAction.prototype.widgetFactory;
    /**
     * @type {?}
     * @private
     */
    FormElementComponentAction.prototype.terminator;
}
