// ngx-schema-form/esm5/lib/property-binding-registry.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/**
 * General purpose propery binding registry
 */
var /**
 * General purpose propery binding registry
 */
PropertyBindingRegistry = /** @class */ (function () {
    function PropertyBindingRegistry() {
        this.bindings = {};
    }
    /**
     * @param {?} type
     * @return {?}
     */
    PropertyBindingRegistry.prototype.getPropertyBindings = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        this.bindings[type] = this.bindings[type] || new PropertyBindings();
        return this.bindings[type];
    };
    /**
     * @return {?}
     */
    PropertyBindingRegistry.prototype.getPropertyBindingsVisibility = /**
     * @return {?}
     */
    function () {
        return this.getPropertyBindings(PropertyBindingTypes.visibility);
    };
    return PropertyBindingRegistry;
}());
/**
 * General purpose propery binding registry
 */
export { PropertyBindingRegistry };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PropertyBindingRegistry.prototype.bindings;
}
/** @enum {number} */
var PropertyBindingTypes = {
    visibility: 0,
};
export { PropertyBindingTypes };
PropertyBindingTypes[PropertyBindingTypes.visibility] = 'visibility';
/**
 * Storage that holds all bindings that are property paths related.<br/>
 */
var /**
 * Storage that holds all bindings that are property paths related.<br/>
 */
PropertyBindings = /** @class */ (function () {
    function PropertyBindings() {
        this.sourcesIndex = new SimplePropertyIndexer();
        this.dependenciesIndex = new SimplePropertyIndexer();
    }
    /**
     * @param {?} dependencyPath
     * @param {?} sourcePropertyPath
     * @return {?}
     */
    PropertyBindings.prototype.add = /**
     * @param {?} dependencyPath
     * @param {?} sourcePropertyPath
     * @return {?}
     */
    function (dependencyPath, sourcePropertyPath) {
        this.sourcesIndex.store(sourcePropertyPath, dependencyPath);
        this.dependenciesIndex.store(dependencyPath, sourcePropertyPath);
    };
    /**
     * @param {?} dependencyPath
     * @return {?}
     */
    PropertyBindings.prototype.findByDependencyPath = /**
     * @param {?} dependencyPath
     * @return {?}
     */
    function (dependencyPath) {
        var e_1, _a;
        /** @type {?} */
        var result = this.dependenciesIndex.find(dependencyPath);
        result.results = result.results || [];
        /** @type {?} */
        var values = [];
        try {
            for (var _b = tslib_1.__values(result.results), _c = _b.next(); !_c.done; _c = _b.next()) {
                var res = _c.value;
                values = values.concat(Object.keys(res.value));
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return result.found ? values : [];
    };
    /**
     * @param {?} sourcePropertyPath
     * @return {?}
     */
    PropertyBindings.prototype.getBySourcePropertyPath = /**
     * @param {?} sourcePropertyPath
     * @return {?}
     */
    function (sourcePropertyPath) {
        var e_2, _a;
        /** @type {?} */
        var result = this.sourcesIndex.find(sourcePropertyPath);
        result.results = result.results || [];
        /** @type {?} */
        var values = [];
        try {
            for (var _b = tslib_1.__values(result.results), _c = _b.next(); !_c.done; _c = _b.next()) {
                var res = _c.value;
                values = values.concat(Object.keys(res.value));
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return result.found ? values : [];
    };
    /**
     * @param {?} path
     * @return {?}
     */
    PropertyBindings.prototype.createPathIndex = /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return path.split('/');
    };
    return PropertyBindings;
}());
/**
 * Storage that holds all bindings that are property paths related.<br/>
 */
export { PropertyBindings };
if (false) {
    /** @type {?} */
    PropertyBindings.prototype.sourcesIndex;
    /** @type {?} */
    PropertyBindings.prototype.dependenciesIndex;
}
/**
 * Simple indexer to store property paths
 */
var SimplePropertyIndexer = /** @class */ (function () {
    function SimplePropertyIndexer() {
        this.index = {};
        this.findOnlyWithValue = true;
    }
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    SimplePropertyIndexer.prototype._createPathIndex = /**
     * @private
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return path
            .replace(new RegExp('//', 'g'), '/')
            .replace(new RegExp('^/', 'g'), '')
            .split('/').filter((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item; }));
    };
    /**
     * @param {?} propertyPath
     * @param {?=} value
     * @return {?}
     */
    SimplePropertyIndexer.prototype.store = /**
     * @param {?} propertyPath
     * @param {?=} value
     * @return {?}
     */
    function (propertyPath, value) {
        this._storeIndex(this._createPathIndex(propertyPath), value);
    };
    /**
     * @private
     * @param {?} pathIndex
     * @param {?=} value
     * @return {?}
     */
    SimplePropertyIndexer.prototype._storeIndex = /**
     * @private
     * @param {?} pathIndex
     * @param {?=} value
     * @return {?}
     */
    function (pathIndex, value) {
        var e_3, _a;
        /** @type {?} */
        var indexPos = this.index;
        try {
            for (var pathIndex_1 = tslib_1.__values(pathIndex), pathIndex_1_1 = pathIndex_1.next(); !pathIndex_1_1.done; pathIndex_1_1 = pathIndex_1.next()) {
                var key = pathIndex_1_1.value;
                indexPos[key] = indexPos[key] || {};
                indexPos = indexPos[key];
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (pathIndex_1_1 && !pathIndex_1_1.done && (_a = pathIndex_1.return)) _a.call(pathIndex_1);
            }
            finally { if (e_3) throw e_3.error; }
        }
        if (indexPos && value) {
            indexPos[SimplePropertyIndexer.MARKER] = indexPos[SimplePropertyIndexer.MARKER] || {};
            indexPos[SimplePropertyIndexer.MARKER][value] = value;
        }
    };
    /**
     * Find path in index.<br/>
     * Will find path like:<br/>
     * <ul>
     *     <li>/property/0/prop</li>
     *     <li>/property/0/prop/2/test</li>
     *     <li>/property/0/prop/&#42;/test</li>
     *     <li>/property/&#42;/prop/1/test</li>
     *     <li>/property/&#42;/prop/&#42;/test</li>
     *     <li>/property/1/prop/&#42;/test</li>
     *  </ul>
     * @param path
     */
    /**
     * Find path in index.<br/>
     * Will find path like:<br/>
     * <ul>
     *     <li>/property/0/prop</li>
     *     <li>/property/0/prop/2/test</li>
     *     <li>/property/0/prop/&#42;/test</li>
     *     <li>/property/&#42;/prop/1/test</li>
     *     <li>/property/&#42;/prop/&#42;/test</li>
     *     <li>/property/1/prop/&#42;/test</li>
     *  </ul>
     * @param {?} path
     * @return {?}
     */
    SimplePropertyIndexer.prototype.find = /**
     * Find path in index.<br/>
     * Will find path like:<br/>
     * <ul>
     *     <li>/property/0/prop</li>
     *     <li>/property/0/prop/2/test</li>
     *     <li>/property/0/prop/&#42;/test</li>
     *     <li>/property/&#42;/prop/1/test</li>
     *     <li>/property/&#42;/prop/&#42;/test</li>
     *     <li>/property/1/prop/&#42;/test</li>
     *  </ul>
     * @param {?} path
     * @return {?}
     */
    function (path) {
        return this._findInIndex(this._createPathIndex(path));
    };
    /**
     * @param {?} path
     * @return {?}
     */
    SimplePropertyIndexer.prototype._findInIndex = /**
     * @param {?} path
     * @return {?}
     */
    function (path) {
        /** @type {?} */
        var ixRes = { target: path, found: false, results: [] };
        this.__findIndex(ixRes, path, this.index, []);
        return ixRes;
    };
    /**
     * @param {?} indexerResults
     * @param {?} path
     * @param {?} index
     * @param {?=} parent
     * @return {?}
     */
    SimplePropertyIndexer.prototype.__findIndex = /**
     * @param {?} indexerResults
     * @param {?} path
     * @param {?} index
     * @param {?=} parent
     * @return {?}
     */
    function (indexerResults, path, index, parent) {
        var e_4, _a;
        /** @type {?} */
        var p = parent || [];
        /** @type {?} */
        var segment = path[0];
        /** @type {?} */
        var wild = ('*' === segment) ? Object.keys(index) : [];
        /** @type {?} */
        var _keys = ((/** @type {?} */ ((Array.isArray(segment) ? segment : [segment])))).concat(wild);
        /** @type {?} */
        var keys = _keys.filter((/**
         * @param {?} item
         * @param {?} pos
         * @return {?}
         */
        function (item, pos) { return '*' !== item && _keys.indexOf(item) === pos; }));
        if (index['*']) {
            keys.push('*');
        }
        /** @type {?} */
        var paths = [];
        try {
            for (var keys_1 = tslib_1.__values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                var key = keys_1_1.value;
                /** @type {?} */
                var restPath = path.slice(1);
                /** @type {?} */
                var restIndex = index[key];
                /** @type {?} */
                var restParent = p.concat(key);
                if (path.length === 1) { // collect only the full paths
                    if (!this.findOnlyWithValue || (restIndex && restIndex[SimplePropertyIndexer.MARKER])) {
                        indexerResults.results = indexerResults.results || [];
                        indexerResults.results.push({
                            path: restParent,
                            value: restIndex[SimplePropertyIndexer.MARKER]
                        });
                        paths.push(restParent);
                        indexerResults.found = indexerResults.results.length > 0;
                    }
                }
                if (!restPath || !restPath.length || !restIndex) {
                    break;
                }
                /** @type {?} */
                var restPaths = this.__findIndex(indexerResults, restPath, restIndex, restParent);
                paths = paths.concat(restPaths);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return paths;
    };
    SimplePropertyIndexer.MARKER = '$____value';
    return SimplePropertyIndexer;
}());
export { SimplePropertyIndexer };
if (false) {
    /** @type {?} */
    SimplePropertyIndexer.MARKER;
    /** @type {?} */
    SimplePropertyIndexer.prototype.index;
    /** @type {?} */
    SimplePropertyIndexer.prototype.findOnlyWithValue;
}
/**
 * @record
 */
export function IndexerResult() { }
if (false) {
    /**
     * The path originally searched for
     * @type {?}
     */
    IndexerResult.prototype.target;
    /**
     * Flag for the status of found or not found.<br/>
     * Usually <code>results</code> will be empty if no matches found.
     * @type {?}
     */
    IndexerResult.prototype.found;
    /**
     * The result path and values from the index search.<br/>
     * Usually <code>results</code> will be empty if no matches found.
     * @type {?}
     */
    IndexerResult.prototype.results;
}
