// ngx-schema-form/esm5/lib/widget.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/**
 * @abstract
 * @template T
 */
var /**
 * @abstract
 * @template T
 */
Widget = /** @class */ (function () {
    function Widget() {
        this.id = '';
        this.name = '';
        this.schema = {};
    }
    return Widget;
}());
/**
 * @abstract
 * @template T
 */
export { Widget };
if (false) {
    /** @type {?} */
    Widget.prototype.formProperty;
    /** @type {?} */
    Widget.prototype.control;
    /** @type {?} */
    Widget.prototype.errorMessages;
    /** @type {?} */
    Widget.prototype.id;
    /** @type {?} */
    Widget.prototype.name;
    /** @type {?} */
    Widget.prototype.schema;
}
var ControlWidget = /** @class */ (function (_super) {
    tslib_1.__extends(ControlWidget, _super);
    function ControlWidget() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    ControlWidget.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var control = this.control;
        this.formProperty.valueChanges.subscribe((/**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            if (control.value !== newValue) {
                control.setValue(newValue, { emitEvent: false });
            }
        }));
        this.formProperty.errorsChanges.subscribe((/**
         * @param {?} errors
         * @return {?}
         */
        function (errors) {
            control.setErrors(errors, { emitEvent: true });
            /** @type {?} */
            var messages = (errors || [])
                .filter((/**
             * @param {?} e
             * @return {?}
             */
            function (e) {
                return e.path && e.path.slice(1) === _this.formProperty.path;
            }))
                .map((/**
             * @param {?} e
             * @return {?}
             */
            function (e) { return e.message; }));
            _this.errorMessages = messages.filter((/**
             * @param {?} m
             * @param {?} i
             * @return {?}
             */
            function (m, i) { return messages.indexOf(m) === i; }));
        }));
        control.valueChanges.subscribe((/**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            _this.formProperty.setValue(newValue, false);
        }));
    };
    return ControlWidget;
}(Widget));
export { ControlWidget };
var ArrayLayoutWidget = /** @class */ (function (_super) {
    tslib_1.__extends(ArrayLayoutWidget, _super);
    function ArrayLayoutWidget() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    ArrayLayoutWidget.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var control = this.control;
        this.formProperty.errorsChanges.subscribe((/**
         * @param {?} errors
         * @return {?}
         */
        function (errors) {
            control.setErrors(errors, { emitEvent: true });
        }));
    };
    return ArrayLayoutWidget;
}(Widget));
export { ArrayLayoutWidget };
var ObjectLayoutWidget = /** @class */ (function (_super) {
    tslib_1.__extends(ObjectLayoutWidget, _super);
    function ObjectLayoutWidget() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    ObjectLayoutWidget.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var control = this.control;
        this.formProperty.errorsChanges.subscribe((/**
         * @param {?} errors
         * @return {?}
         */
        function (errors) {
            control.setErrors(errors, { emitEvent: true });
        }));
    };
    return ObjectLayoutWidget;
}(Widget));
export { ObjectLayoutWidget };
