// ngx-schema-form/esm5/lib/widgetchooser.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ChangeDetectorRef, EventEmitter, Input, Output, ViewChild, ViewContainerRef } from '@angular/core';
import { TerminatorService } from './terminator.service';
import { WidgetFactory } from './widgetfactory';
var WidgetChooserComponent = /** @class */ (function () {
    function WidgetChooserComponent(widgetFactory, cdr, terminator) {
        if (widgetFactory === void 0) { widgetFactory = null; }
        this.widgetFactory = widgetFactory;
        this.cdr = cdr;
        this.terminator = terminator;
        this.widgetInstanciated = new EventEmitter();
    }
    /**
     * @return {?}
     */
    WidgetChooserComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.subs = this.terminator.onDestroy.subscribe((/**
         * @param {?} destroy
         * @return {?}
         */
        function (destroy) {
            if (destroy) {
                _this.ref.destroy();
            }
        }));
    };
    /**
     * @return {?}
     */
    WidgetChooserComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.ref = this.widgetFactory.createWidget(this.container, this.widgetInfo.id);
        this.widgetInstanciated.emit(this.ref.instance);
        this.widgetInstance = this.ref.instance;
        this.cdr.detectChanges();
    };
    /**
     * @return {?}
     */
    WidgetChooserComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.subs.unsubscribe();
    };
    WidgetChooserComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sf-widget-chooser',
                    template: "<div #target></div>"
                }] }
    ];
    /** @nocollapse */
    WidgetChooserComponent.ctorParameters = function () { return [
        { type: WidgetFactory },
        { type: ChangeDetectorRef },
        { type: TerminatorService }
    ]; };
    WidgetChooserComponent.propDecorators = {
        widgetInfo: [{ type: Input }],
        widgetInstanciated: [{ type: Output }],
        container: [{ type: ViewChild, args: ['target', { read: ViewContainerRef, static: true },] }]
    };
    return WidgetChooserComponent;
}());
export { WidgetChooserComponent };
if (false) {
    /** @type {?} */
    WidgetChooserComponent.prototype.widgetInfo;
    /** @type {?} */
    WidgetChooserComponent.prototype.widgetInstanciated;
    /** @type {?} */
    WidgetChooserComponent.prototype.container;
    /**
     * @type {?}
     * @private
     */
    WidgetChooserComponent.prototype.widgetInstance;
    /**
     * @type {?}
     * @private
     */
    WidgetChooserComponent.prototype.ref;
    /**
     * @type {?}
     * @private
     */
    WidgetChooserComponent.prototype.subs;
    /**
     * @type {?}
     * @private
     */
    WidgetChooserComponent.prototype.widgetFactory;
    /**
     * @type {?}
     * @private
     */
    WidgetChooserComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    WidgetChooserComponent.prototype.terminator;
}
