// ngx-schema-form/esm5/lib/template-schema/field/field.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ContentChildren, QueryList, ElementRef } from '@angular/core';
import { merge } from 'rxjs';
import { ActionRegistry } from '../../model/actionregistry';
import { TemplateSchemaService } from '../template-schema.service';
import { ButtonComponent } from '../button/button.component';
import { FieldParent } from './field-parent';
import { FieldType } from './field';
import { ItemComponent } from './item/item.component';
var FieldComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FieldComponent, _super);
    function FieldComponent(elementRef, templateSchemaService, actionRegistry) {
        var _this = _super.call(this) || this;
        _this.elementRef = elementRef;
        _this.templateSchemaService = templateSchemaService;
        _this.actionRegistry = actionRegistry;
        _this.type = FieldType.String;
        _this.schema = {};
        return _this;
    }
    /**
     * @return {?}
     */
    FieldComponent.prototype.getSchema = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var _a = this.getFieldsSchema(this.childFields.filter((/**
         * @param {?} field
         * @return {?}
         */
        function (field) { return field !== _this; }))), properties = _a.properties, items = _a.items, required = _a.required;
        /** @type {?} */
        var oneOf = this.getOneOf();
        /** @type {?} */
        var schema = (/** @type {?} */ ({
            type: this.type
        }));
        if (this.title !== undefined) {
            schema.title = this.title;
        }
        if (properties !== undefined) {
            schema.properties = properties;
        }
        if (items !== undefined) {
            schema.items = items;
        }
        // requried child fields
        if (required !== undefined) {
            schema.required = required;
        }
        if (oneOf !== undefined) {
            schema.oneOf = oneOf;
        }
        if (this.description !== undefined) {
            schema.description = this.description;
        }
        if (this.placeholder !== undefined) {
            schema.placeholder = this.placeholder;
        }
        if (this.format !== undefined) {
            schema.format = this.format;
        }
        if (this.widget !== undefined) {
            schema.widget = this.widget;
        }
        if (this.readOnly !== undefined) {
            schema.readOnly = this.readOnly;
        }
        /** @type {?} */
        var buttons = this.getButtons();
        if (buttons.length > 0) {
            schema.buttons = buttons;
        }
        // @Input schema takes precedence
        return Object.assign(schema, this.schema);
    };
    /**
     * @return {?}
     */
    FieldComponent.prototype.getValidators = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // registering validator here is not possible since prop full path is needed
        /** @type {?} */
        var childValidators = this.getFieldsValidators(this.childFields.filter((/**
         * @param {?} field
         * @return {?}
         */
        function (field) { return field !== _this; })));
        /** @type {?} */
        var validators = childValidators.map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var path = _a.path, validator = _a.validator;
            return {
                path: _this.path + path,
                validator: validator
            };
        }));
        if (!this.validator) {
            return validators;
        }
        validators.push({ path: this.path, validator: this.validator });
        return validators;
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    FieldComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var e_1, _a;
        /** @type {?} */
        var keys = Object.keys(changes);
        if (keys.length > 0) {
            try {
                for (var keys_1 = tslib_1.__values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                    var key = keys_1_1.value;
                    if (!changes[key].isFirstChange()) {
                        // on any input change, force schema change generation
                        this.templateSchemaService.changed();
                        break;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
    };
    /**
     * @private
     * @return {?}
     */
    FieldComponent.prototype.getOneOf = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.childItems.length === 0) {
            return;
        }
        /** @type {?} */
        var items = this.childItems.map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var value = _a.value, description = _a.description;
            if (!Array.isArray(value)) {
                return { enum: [value], description: description };
            }
            return { enum: value, description: description };
        }));
        if (items.length === 0) {
            return;
        }
        return items;
    };
    /**
     * @private
     * @return {?}
     */
    FieldComponent.prototype.setTitleFromContent = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var textContent = this.getTextContent(this.elementRef);
        //  title as @Input takes priority over content text
        if (textContent && !this.title) {
            this.title = textContent;
        }
    };
    /**
     * @return {?}
     */
    FieldComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // cache it
        this.setTitleFromContent();
        merge(this.childFields.changes, this.childItems.changes, this.childButtons.changes)
            .subscribe((/**
         * @return {?}
         */
        function () { return _this.templateSchemaService.changed(); }));
    };
    FieldComponent.decorators = [
        { type: Component, args: [{
                    selector: 'sf-field',
                    template: "<ng-content ></ng-content>\n"
                }] }
    ];
    /** @nocollapse */
    FieldComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: TemplateSchemaService },
        { type: ActionRegistry }
    ]; };
    FieldComponent.propDecorators = {
        childFields: [{ type: ContentChildren, args: [FieldComponent,] }],
        childItems: [{ type: ContentChildren, args: [ItemComponent,] }],
        childButtons: [{ type: ContentChildren, args: [ButtonComponent,] }],
        name: [{ type: Input }],
        type: [{ type: Input }],
        format: [{ type: Input }],
        required: [{ type: Input }],
        readOnly: [{ type: Input }],
        title: [{ type: Input }],
        description: [{ type: Input }],
        placeholder: [{ type: Input }],
        widget: [{ type: Input }],
        validator: [{ type: Input }],
        schema: [{ type: Input }]
    };
    return FieldComponent;
}(FieldParent));
export { FieldComponent };
if (false) {
    /** @type {?} */
    FieldComponent.prototype.childFields;
    /** @type {?} */
    FieldComponent.prototype.childItems;
    /** @type {?} */
    FieldComponent.prototype.childButtons;
    /** @type {?} */
    FieldComponent.prototype.name;
    /** @type {?} */
    FieldComponent.prototype.type;
    /** @type {?} */
    FieldComponent.prototype.format;
    /** @type {?} */
    FieldComponent.prototype.required;
    /** @type {?} */
    FieldComponent.prototype.readOnly;
    /** @type {?} */
    FieldComponent.prototype.title;
    /** @type {?} */
    FieldComponent.prototype.description;
    /** @type {?} */
    FieldComponent.prototype.placeholder;
    /** @type {?} */
    FieldComponent.prototype.widget;
    /** @type {?} */
    FieldComponent.prototype.validator;
    /** @type {?} */
    FieldComponent.prototype.schema;
    /**
     * @type {?}
     * @private
     */
    FieldComponent.prototype.elementRef;
    /**
     * @type {?}
     * @private
     */
    FieldComponent.prototype.templateSchemaService;
    /**
     * @type {?}
     * @protected
     */
    FieldComponent.prototype.actionRegistry;
}
